src/viewer.js

```diff
@@ -50,18 +49,34 @@ class Viewer {
   /**
    * @param {!Function} showViewer method that shows the viewer.
    * @param {!Function} hideViewer method that hides the viewer.
+   * @param {!Function():boolean} isViewerHidden method that determines if viewer is hidden.
    */
-  setViewerShowAndHide(showViewer, hideViewer) {
+  setViewerShowAndHide(showViewer, hideViewer, isViewerHidden) {
     /** @private {!Function} */
     this.showViewer_ = showViewer;
     /** @private {!Function} */
     this.hideViewer_ = hideViewer;
+    /** @private {!Function} */
+    this.isViewerHidden_ = isViewerHidden;
+  }
+
+  /**
+   * @return {boolean} true if the viewer has already been loaded.
+   * @private
+   */
+  isLoaded_() {
+    return !!this.iframe_ && !!this.viewerMessaging_;
   }
 
   /**
    * Attaches the AMP Doc Iframe to the Host Element.
    */
   attach() {
+    if (this.isLoaded_()) {
+      this.history_.goForward();
```

Comment:
I thought I had a comment about this before, but don't see it now. I think you don't want to depend on what's on the forward history stack when you're reattching. If something else on the page has manipulated  the history, it could be totally unrelated to AMP.

Sorry if my previous comment mislead you. I think what you had initially seemed reasonable.